Frameworks/Foundation/NSURLConnection.mm

```diff
@@ -19,168 +20,289 @@
+@end
 
-    id response, error;
-    id data = [self sendSynchronousRequest:request returningResponse:&response error:&error];
+@implementation _NSURLConnection_DataDelegate {
+@private
+    StrongId<NSOperationQueue> _operationQueue;
+    StrongId<void (^)(NSURLResponse*, NSData*, NSError*)> _completionHandler;
 
-    completionHandler(response, data, error);
+    StrongId<NSURLResponse> _response;
+    StrongId<NSMutableData> _data;
 }
 
-/**
- @Status Caveat
- @Notes NSError returned is not detailed
-*/
-+ (NSData*)sendSynchronousRequest:(id)request returningResponse:(NSURLResponse**)responsep error:(NSError**)errorp {
-    NSURLConnectionState* state = [[[NSURLConnectionState alloc] init] autorelease];
-    NSURLConnection* connection = [[self alloc] initWithRequest:request delegate:state startImmediately:FALSE];
++ (_NSURLConnection_DataDelegate*)delegateWithQueue:(NSOperationQueue*)queue handler:(void (^)(NSURLResponse*, NSData*, NSError*))handler {
+    _NSURLConnection_DataDelegate* ret = [[_NSURLConnection_DataDelegate new] autorelease];
+    if (ret) {
+        ret->_operationQueue = queue;
+        ret->_completionHandler = handler;
+    }
+    return ret;
+}
 
-    if (connection == nil) {
-        if (errorp) {
-            *errorp = [NSError errorWithDomain:@"NSURLErrorDomain" code:50 userInfo:nil];
-        }
+- (void)connection:(NSURLConnection*)connection didFailWithError:(NSError*)error {
+    [_operationQueue addOperation:[NSBlockOperation blockOperationWithBlock:^void() {
+                         _completionHandler(_response.get(), nil, error);
+                     }]];
+}
 
-        return nil;
+- (void)connection:(NSURLConnection*)connection didReceiveResponse:(NSURLResponse*)response {
+    // We get some of these for each redirect. Only capture the final one.
+    _response = response;
+    if (!_data.get() || [_data length] > 0) {
+        _data.attach([NSMutableData new]);
     }
+}
 
-    id mode = @"NSURLConnectionRequestMode";
+- (void)connection:(NSURLConnection*)connection didReceiveData:(NSData*)data {
+    [_data appendData:data];
+}
 
-    [connection scheduleInRunLoop:[NSRunLoop currentRunLoop] forMode:mode];
-    [connection start];
+- (void)connectionDidFinishLoading:(NSURLConnection*)connection {
+    [_operationQueue addOperation:[NSBlockOperation blockOperationWithBlock:^void() {
+                         _completionHandler(_response.get(), _data.get(), nil);
+                     }]];
+}
 
-    [state receiveAllDataInMode:mode];
-    [connection unscheduleFromRunLoop:[NSRunLoop currentRunLoop] forMode:mode];
+@end
 
-    id result = state.receivedData;
+// Actual NSURLConnection implementation
+@implementation NSURLConnection {
+@private
+    StrongId<NSObject> _delegate;
 
-    [connection cancel];
+    std::vector<std::pair<StrongId<NSRunLoop>, StrongId<NSString>>> _scheduledRunLoops; // NSRunLoop, NSRunLoopMode
+    // TODO #: Cannot currently schedule on more than one runloop
```

Comment:
i know, i knowww - i dont like filing these tasks until the PR is actually out because they often lack context without the PR,